bin/i18n/curriculum_sync_utils/serializers.rb

```diff
@@ -0,0 +1,110 @@
+module CurriculumSyncUtils
+  # Define a custom CollectionSerializer subclass, optimized for crowdin compatibility.
+  class CrowdinCollectionSerializer < ActiveModel::Serializer::CollectionSerializer
+    # @override
+    def serializable_hash(adapter_options, options, adapter_instance)
+      # Default CollectionSerializer behavior is to return an array of hashes,
+      # where each hash is a single result. We would instead like to return a
+      # single hash, keyed by `crowdin_key`
+      return super.reduce({}) do |results, result|
+        raise KeyEror.new("Serializer must define :crowdin_key for curriculum content I18N serialization; got #{result.keys.inspect}") unless result.key?(:crowdin_key)
+        crowdin_key = result.delete(:crowdin_key)
+        results[crowdin_key] = result unless result.empty?
+        results
+      end
+    end
+  end
+
+  # Define a custom Serializer subclass, optimized for crowdin compatibility.
+  class CrowdinSerializer < ActiveModel::Serializer
+    config.collection_serializer = CrowdinCollectionSerializer
+    config.adapter = :json
+
+    # @override
+    def serializable_hash(adapter_options = nil, options = {}, adapter_instance = self.class.serialization_adapter_instance)
+      # include nested relationships of an arbitrary depth, rather than the
+      # default of only a single level of nesting.
+      options[:include_directive] ||= JSONAPI::IncludeDirective.new('**', allow_wildcard: true)
+      # compact the result, excluding not only nil values but also empty ones
+      super.reject {|_, v| v.blank?}
+    end
+
+    # Add a 'crowdin_key' attribute to every model. Should return a value that
+    # is both capable of uniquely identifying the object in question, and is
+    # useful for translators to see.
+    #
+    # Note that any subclass which overrides this method will also need some
+    # associated postprocessing in the sync out.
+    #
+    # Required for compatiblity with CrowdinCollectionSerializer
+    attribute :crowdin_key
+    def crowdin_key
+      object.try(:key) || object.id
```

Comment:
Some of our models have keys that are only unique within their parent (lesson or course version). I know vocab and resources do this at least. Should we schedule some work to add uniquely identifiable keys to these objects?

Reply:
oh nevermind, it looks like the way that the data is nested makes this okay. I can't think of anything that wouldn't have a unique key within a lesson

Reply:
oh, actually having globally-unique keys will be necessary for the rendering of these strings.

In general, we should always have a way of uniquely identifying any piece of content in our system in an environment-agnostic way. So I'd say that yes, we should schedule some work to add that. FYI @tess323

Reply:
To be clear, we can easily create a globally unique key, even if the strings are flattened. For example, for resources, we'd identify them similarly to how we do for the markdown syntax I'd imagine